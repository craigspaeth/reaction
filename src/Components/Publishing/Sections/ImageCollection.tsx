import { find } from "lodash"
import * as React from "react"
import sizeMe from "react-sizeme"
import styled from "styled-components"
import fillwidthDimensions from "../../../Utils/fillwidth"
import { pMedia } from "../../Helpers"
import { sizeMeRefreshRate } from "../Constants"
import { SectionLayout } from "../Typings"
import { Artwork } from "./Artwork"
import { Image } from "./Image"
import { ImageCollectionItem } from "./ImageCollectionItem"

interface ImageCollectionProps {
  images: any
  targetHeight?: number
  gutter?: number
  sectionLayout?: SectionLayout
  size?: {
    width: number
  }
}

const sizeMeOptions = {
  refreshRate: sizeMeRefreshRate,
  noPlaceholder: true,
}

const ImageCollectionComponent: React.SFC<ImageCollectionProps> = props => {
  const {
    gutter,
    images,
    sectionLayout,
    size,
    targetHeight,
  } = props

  const dimensions = fillwidthDimensions(images, size.width, gutter, targetHeight)
  const renderedImages = renderImages(images, dimensions, gutter, sectionLayout, size.width)

  return (
    <ImageCollectionContainer>
      {renderedImages}
    </ImageCollectionContainer>
  )
}

ImageCollectionComponent.defaultProps = {
  size: {
    width: 680,
  },
}

function renderImages(images, dimensions, gutter, sectionLayout, width) {
  const renderedImages = images.map((image, i) => {
    const url = image.url || image.image

    let imageSize
    if (width <= 600 || dimensions.length === 1) {
      imageSize = {}
    } else {
      imageSize = find(dimensions, ["__id", url])
    }

    let renderedImage
    if (image.type === "image") {
      renderedImage = (
        <Image
          image={image}
          sectionLayout={sectionLayout}
          width={imageSize.width}
          height={imageSize.height}
        />
      )
    } else if (image.type === "artwork") {
      renderedImage = (
        <Artwork
          artwork={image}
          sectionLayout={sectionLayout}
          width={imageSize.width}
          height={imageSize.height}
        />
      )
    } else {
      return false
    }

    const margin = i === dimensions.length - 1 ? 0 : gutter

    return (
      <ImageCollectionItem
        key={i}
        margin={margin}
        width={imageSize.width}
      >
        {renderedImage}
      </ImageCollectionItem>
    )
  })
  return renderedImages
}

const ImageCollectionContainer = styled.div`
  display: flex;
  width: 100%;

  // TODO: Follow up on https://github.com/artsy/publishing/issues/80#issuecomment-338055364
  justify-content: center;

  ${pMedia.xs`
    flex-direction: column;
  `};
`

export const ImageCollection = sizeMe(sizeMeOptions)(ImageCollectionComponent)
