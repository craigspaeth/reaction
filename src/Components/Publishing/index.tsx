// Components
export { Article } from "./Article"
export { Header } from "./Header/Header"
export { RelatedArticlesCanvas } from "./RelatedArticles/RelatedArticlesCanvas"
export { RelatedArticlesPanel } from "./RelatedArticles/RelatedArticlesPanel"
export { Artwork } from "./Sections/Artwork"
export { Authors } from "./Sections/Authors"
export { Caption } from "./Sections/Caption"
export { Embed } from "./Sections/Embed"
export { FullscreenViewer } from "./Sections/FullscreenViewer/FullscreenViewer"
export { Image } from "./Sections/Image"
export { ImageCollection } from "./Sections/ImageCollection"
export { ImageSetPreview } from "./Sections/ImageSetPreview"
export { ImageSetPreviewClassic } from "./Sections/ImageSetPreviewClassic"
export { Text } from "./Sections/Text"
export { Video } from "./Sections/Video"

// Icon SVGs
export { IconEditEmbed } from "./Icon/IconEditEmbed"
export { IconEditImages } from "./Icon/IconEditImages"
export { IconEditSection } from "./Icon/IconEditSection"
export { IconEditText } from "./Icon/IconEditText"
export { IconEditVideo } from "./Icon/IconEditVideo"
export { IconHeroImage } from "./Icon/IconHeroImage"
export { IconHeroVideo } from "./Icon/IconHeroVideo"
export { IconImageFullscreen } from "./Icon/IconImageFullscreen"
export { IconImageSet } from "./Icon/IconImageSet"
export { IconLayoutFullscreen } from "./Icon/IconLayoutFullscreen"
export { IconLayoutSplit } from "./Icon/IconLayoutSplit"
export { IconLayoutText } from "./Icon/IconLayoutText"
export { IconRemove } from "./Icon/IconRemove"

// Test Fixtures
import {
  ClassicArticle,
  FeatureArticle,
  ImageHeavyStandardArticle,
  ShortStandardArticle,
  StandardArticle,
} from "./Fixtures/Articles"

// FIXME: Refactor out SizeMe; see https://github.com/ctrlplusb/react-sizeme#server-side-rendering
import sizeMe from "react-sizeme"
sizeMe.noPlaceholders = true

export const Fixtures = {
  ClassicArticle,
  FeatureArticle,
  ImageHeavyStandardArticle,
  ShortStandardArticle,
  StandardArticle,
}


// Constants
import * as AllConstants from "./Constants"
export const Constants = AllConstants
