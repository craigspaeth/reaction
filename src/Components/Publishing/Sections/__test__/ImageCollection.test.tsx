import "jest-styled-components"
import * as React from "react"
import * as renderer from "react-test-renderer"
import { Images } from "../../Fixtures/Components"
import { ImageCollection } from "../ImageCollection"

jest.mock("react-sizeme", () => jest.fn(c => d => d))
jest.mock("react-lines-ellipsis/lib/html", () => {
  const React = require('react')
  return () => <div />
})

jest.mock('react-dom/server', () => ({
  renderToStaticMarkup: (x) => x
}))

it("renders properly", () => {
  const image = renderer.create(<ImageCollection images={Images} targetHeight={400} gutter={10} />).toJSON()
  expect(image).toMatchSnapshot()
})

it("renders a single image properly", () => {
  const image = renderer.create(<ImageCollection images={[Images[0]]} targetHeight={400} gutter={10} />).toJSON()
  expect(image).toMatchSnapshot()
})
