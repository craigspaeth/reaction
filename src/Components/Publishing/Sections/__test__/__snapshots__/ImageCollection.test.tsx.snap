// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a single image properly 1`] = `
.c8 {
  padding: 0;
  margin-top: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: Unica77LLWebRegular , 'Arial', 'serif';
  -webkit-font-smoothing: antialiased;
  font-size: 14px;
  line-height: 1.4em;
}

.c8 .title {
  font-family: Unica77LLWebItalic , 'Arial', 'serif';
  -webkit-font-smoothing: antialiased;
  font-size: 14px;
  line-height: 1.4em;
}

.c9 {
  color: #999;
}

.c9 .name:after {
  content: ', ';
}

.c9 .title:after {
  content: ', ';
}

.c9 .date:after {
  content: ', ';
}

.c7 {
  position: absolute;
  bottom: 0;
  right: 0;
  margin: 10px;
  width: 25px;
  height: 25px;
  cursor: pointer;
  opacity: 0.6;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c7:hover {
  opacity: 1;
}

.c6 {
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c3 {
  position: relative;
}

.c3:hover .c5 {
  opacity: 1;
}

.c4 {
  display: block;
}

.c2 {
  text-decoration: none;
}

.c1 {
  margin-right: 0px;
  width: 100%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

@media (max-width:600px) {
  .c8 {
    padding: 0 10px;
  }
}

@media (max-width:720px) {
  .c3 .c5 {
    opacity: 1;
  }
}

@media (max-width:600px) {
  .c1 {
    margin-bottom: 10px;
  }
}

@media (max-width:600px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
    width={undefined}
  >
    <div
      className="display-artwork"
    >
      <a
        className="c2"
        href="/artwork/fernando-botero-nude-on-the-beach"
      >
        <div
          className="c3"
        >
          <img
            alt="Nude on the Beach"
            className="display-artwork__image c4"
            height="auto"
            src="https://d7hftxdivxxvm.cloudfront.net?resize_to=width&src=https%3A%2F%2Fd32dm0rphc51dk.cloudfront.net%2F0aRUvnVgQKbQk5dj8xcCAg%2Flarger.jpg&width=1200&quality=95"
            width="100%"
          />
          <div
            className="c5 c6"
          >
            <div
              className="c7"
              onClick={[Function]}
            >
              <svg
                className="expand"
                version="1.1"
                viewBox="0 0 27 27"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  fill="none"
                  fillRule="evenodd"
                  stroke="none"
                  strokeWidth="1"
                >
                  <g>
                    <circle
                      cx="13.9575195"
                      cy="13.0588379"
                      fill="#000000"
                      fillRule="nonzero"
                      r="13"
                    />
                    <g
                      stroke="#FFFFFF"
                      transform="translate(8.000000, 7.000000)"
                    >
                      <g>
                        <g
                          transform="translate(0.382532, 0.382532)"
                        >
                          <polyline
                            points="4.78165571 0 11.1571966 0 11.1571966 6.37554094"
                          />
                          <polyline
                            points="-1.01881144e-06 4.78165537 6.37553992 4.78165537 6.37553992 11.1571963"
                            transform="translate(3.187769, 7.969426) rotate(-180.000000) translate(-3.187769, -7.969426) "
                          />
                          <path
                            d="M10.360254,0.796942618 L0.796942618,10.360254"
                            strokeLinecap="square"
                          />
                        </g>
                      </g>
                    </g>
                  </g>
                </g>
              </svg>
            </div>
          </div>
        </div>
      </a>
      <div
        className="display-artwork__caption c8"
      >
        <div
          className="c9"
        >
          <div />
        </div>
      </div>
       */
    </div>
  </div>
</div>
`;

exports[`renders properly 1`] = `
.c8 {
  padding: 0;
  margin-top: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: Unica77LLWebRegular , 'Arial', 'serif';
  -webkit-font-smoothing: antialiased;
  font-size: 14px;
  line-height: 1.4em;
}

.c8 .title {
  font-family: Unica77LLWebItalic , 'Arial', 'serif';
  -webkit-font-smoothing: antialiased;
  font-size: 14px;
  line-height: 1.4em;
}

.c9 {
  color: #999;
}

.c9 .name:after {
  content: ', ';
}

.c9 .title:after {
  content: ', ';
}

.c9 .date:after {
  content: ', ';
}

.c7 {
  position: absolute;
  bottom: 0;
  right: 0;
  margin: 10px;
  width: 25px;
  height: 25px;
  cursor: pointer;
  opacity: 0.6;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c7:hover {
  opacity: 1;
}

.c6 {
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c3 {
  position: relative;
}

.c3:hover .c5 {
  opacity: 1;
}

.c4 {
  display: block;
}

.c2 {
  text-decoration: none;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 10px 0 10px 0;
}

.c12 {
  padding: 0;
  width: 100%;
  word-break: break-word;
}

.c12 > p,
.c12 p,
.c12 .public-DraftEditorPlaceholder-root {
  font-family: Unica77LLWebRegular , 'Arial', 'serif';
  -webkit-font-smoothing: antialiased;
  font-size: 14px;
  line-height: 1.4em;
  color: #999;
  margin: 0;
}

.c12 > a,
.c12 a {
  color: #999;
}

.c12 > a:hover,
.c12 a:hover {
  color: black;
}

.c1 {
  margin-right: 10px;
  width: 312px;
}

.c10 {
  margin-right: 10px;
  width: 168px;
}

.c13 {
  margin-right: 0px;
  width: 178px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

@media (max-width:600px) {
  .c8 {
    padding: 0 10px;
  }
}

@media (max-width:720px) {
  .c3 .c5 {
    opacity: 1;
  }
}

@media (max-width:600px) {
  .c11 {
    padding: 0px 10px;
  }
}

@media (max-width:600px) {
  .c12 {
    padding: 0px;
  }
}

@media (max-width:600px) {
  .c1 {
    margin-bottom: 10px;
  }
}

@media (max-width:600px) {
  .c10 {
    margin-bottom: 10px;
  }
}

@media (max-width:600px) {
  .c13 {
    margin-bottom: 10px;
  }
}

@media (max-width:600px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
    width={312}
  >
    <div
      className="display-artwork"
    >
      <a
        className="c2"
        href="/artwork/fernando-botero-nude-on-the-beach"
      >
        <div
          className="c3"
        >
          <img
            alt="Nude on the Beach"
            className="display-artwork__image c4"
            height={224}
            src="https://d7hftxdivxxvm.cloudfront.net?resize_to=width&src=https%3A%2F%2Fd32dm0rphc51dk.cloudfront.net%2F0aRUvnVgQKbQk5dj8xcCAg%2Flarger.jpg&width=1200&quality=95"
            width={312}
          />
          <div
            className="c5 c6"
          >
            <div
              className="c7"
              onClick={[Function]}
            >
              <svg
                className="expand"
                version="1.1"
                viewBox="0 0 27 27"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  fill="none"
                  fillRule="evenodd"
                  stroke="none"
                  strokeWidth="1"
                >
                  <g>
                    <circle
                      cx="13.9575195"
                      cy="13.0588379"
                      fill="#000000"
                      fillRule="nonzero"
                      r="13"
                    />
                    <g
                      stroke="#FFFFFF"
                      transform="translate(8.000000, 7.000000)"
                    >
                      <g>
                        <g
                          transform="translate(0.382532, 0.382532)"
                        >
                          <polyline
                            points="4.78165571 0 11.1571966 0 11.1571966 6.37554094"
                          />
                          <polyline
                            points="-1.01881144e-06 4.78165537 6.37553992 4.78165537 6.37553992 11.1571963"
                            transform="translate(3.187769, 7.969426) rotate(-180.000000) translate(-3.187769, -7.969426) "
                          />
                          <path
                            d="M10.360254,0.796942618 L0.796942618,10.360254"
                            strokeLinecap="square"
                          />
                        </g>
                      </g>
                    </g>
                  </g>
                </g>
              </svg>
            </div>
          </div>
        </div>
      </a>
      <div
        className="display-artwork__caption c8"
      >
        <div
          className="c9"
        >
          <div />
        </div>
      </div>
       */
    </div>
  </div>
  <div
    className="c10"
    width={168}
  >
    <div
      className="article-image"
    >
      <div
        className="c3"
      >
        <img
          alt="Photo by Adam Kuehl for Artsy."
          className="c4"
          height={224}
          src="https://d7hftxdivxxvm.cloudfront.net?resize_to=width&src=https%3A%2F%2Fartsy-media-uploads.s3.amazonaws.com%2Fco8j2xq40ROMyBrJQm_4eQ%252FDafenOilPaintingVillage_AK03.jpg&width=1200&quality=95"
          width={168}
        />
        <div
          className="c5 c6"
        >
          <div
            className="c7"
            onClick={[Function]}
          >
            <svg
              className="expand"
              version="1.1"
              viewBox="0 0 27 27"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fillRule="evenodd"
                stroke="none"
                strokeWidth="1"
              >
                <g>
                  <circle
                    cx="13.9575195"
                    cy="13.0588379"
                    fill="#000000"
                    fillRule="nonzero"
                    r="13"
                  />
                  <g
                    stroke="#FFFFFF"
                    transform="translate(8.000000, 7.000000)"
                  >
                    <g>
                      <g
                        transform="translate(0.382532, 0.382532)"
                      >
                        <polyline
                          points="4.78165571 0 11.1571966 0 11.1571966 6.37554094"
                        />
                        <polyline
                          points="-1.01881144e-06 4.78165537 6.37553992 4.78165537 6.37553992 11.1571963"
                          transform="translate(3.187769, 7.969426) rotate(-180.000000) translate(-3.187769, -7.969426) "
                        />
                        <path
                          d="M10.360254,0.796942618 L0.796942618,10.360254"
                          strokeLinecap="square"
                        />
                      </g>
                    </g>
                  </g>
                </g>
              </g>
            </svg>
          </div>
        </div>
      </div>
      <div
        className="c11"
      >
        <div
          className="c12"
        >
          <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "<p>Photo by Adam Kuehl for Artsy.</p>",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="c13"
    width={178}
  >
    <div
      className="article-image"
    >
      <div
        className="c3"
      >
        <img
          alt="Photo by Adam Kuehl for Artsy. Image courtesy of the Guggenheim Museum."
          className="c4"
          height={224}
          src="https://d7hftxdivxxvm.cloudfront.net?resize_to=width&src=https%3A%2F%2Fd32dm0rphc51dk.cloudfront.net%2FCpHY-DRr7KW0HGXLslCXHw%2Flarger.jpg&width=1200&quality=95"
          width={178}
        />
        <div
          className="c5 c6"
        >
          <div
            className="c7"
            onClick={[Function]}
          >
            <svg
              className="expand"
              version="1.1"
              viewBox="0 0 27 27"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fillRule="evenodd"
                stroke="none"
                strokeWidth="1"
              >
                <g>
                  <circle
                    cx="13.9575195"
                    cy="13.0588379"
                    fill="#000000"
                    fillRule="nonzero"
                    r="13"
                  />
                  <g
                    stroke="#FFFFFF"
                    transform="translate(8.000000, 7.000000)"
                  >
                    <g>
                      <g
                        transform="translate(0.382532, 0.382532)"
                      >
                        <polyline
                          points="4.78165571 0 11.1571966 0 11.1571966 6.37554094"
                        />
                        <polyline
                          points="-1.01881144e-06 4.78165537 6.37553992 4.78165537 6.37553992 11.1571963"
                          transform="translate(3.187769, 7.969426) rotate(-180.000000) translate(-3.187769, -7.969426) "
                        />
                        <path
                          d="M10.360254,0.796942618 L0.796942618,10.360254"
                          strokeLinecap="square"
                        />
                      </g>
                    </g>
                  </g>
                </g>
              </g>
            </svg>
          </div>
        </div>
      </div>
      <div
        className="c11"
      >
        <div
          className="c12"
        >
          <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "<p>Photo by <a href='artsy.net'>Adam Kuehl</a> for Artsy. Image courtesy of the Guggenheim Museum.</p>",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
